### PR TITLE
front: fix and improve op search logic & results

### LINKED ISSUE #6709: Refacto UI & logic of operational points search
Operational points search is wrongly named "station search" and logic is broken in sorting between ch code and names.

UI is not well adapted, it uses "stationCard" from opendata import, taking too much vertical space.

It lacks something to differentiate main operational points from the other ones.

Here's a new proposal of the same component:
image

### PR DESCRIPTION
**Pull request by nicolaswurtz**
closes #6709 

This PR gives better search user experience of operational points & lines: 
- code is half long, much simple and easier to understand — all logic is at the same place
- it adds an `or` logic in search payload in order to have _all-in-one_ call for all flavour of search (trigram, uic, name)
- ch code search works with part of code
- uic search works with part of number
- line code search works with part of line code
- lines are alphabetically sorted 
- reset button removed & refacto: any click on map removes search markers
- better ui, smaller cards, differentiation between main operational points and the other ones

image

### CODE REVIEW

File: front/src/common/Map/Search/MapSearchOperationalPoint.tsx [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        ch: mainOperationalPointsCHCodes.includes(result.ch) ? '' : result.ch,
+      }))
+      // Begin to filter with main operational points (CH code = ''), if not checked, filter on chCode input field
+      .filter((result) => {
+        if (mainOperationalPointsOnly) return result.ch === '';
+        return chCodeFilter !== ''
+          ? chCodeFilter.trim().toLowerCase() === result.ch.toLocaleLowerCase()
+          : true;
+      })
+      .sort((a, b) => a.name.localeCompare(b.name) || a.ch.localeCompare(b.ch));
```
Comment (SharglutDev) [inline front/src/common/Map/Search/MapSearchOperationalPoint.tsx]: We could avoid the `orderedResults` useState and just put this in a useMemo with the same dependency array as the useEffect line 97.

File: front/src/common/Map/Search/MapSearchOperationalPoint.tsx [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            >
+              <span className="trigram">{searchResult.trigram}</span>
+              <span className="name">
+                {searchResult.name}
+                <span className="ch">{searchResult.ch}</span>
+              </span>
+              <span className="uic">{searchResult.uic}</span>
+            </button>
+          ))}
+      </div>
```
Comment (SharglutDev) [inline front/src/common/Map/Search/MapSearchOperationalPoint.tsx]: To split component responsibility (and reduce component size), I think we should still keep another component who could be called  `<MapSearchResults />` that would be only responsible to display the results list. When mapsearchline and mapsearchsignal will be fixed, we might be able to reuse it for them as well.

Good practice would be also to have a `<OperationalPointCard />` component with just the <button> logic inside it.

The OnResultClick() function could be moved inside it (or inside MapSearchResults if we don't use OperationalPointCard)

File: front/src/common/Map/utils.ts [hunk 3]
```diff
@@ -26,10 +26,10 @@ export const onResultSearchClick = ({
   map,
   updateExtViewport,
   dispatch,
-  setSearch,
+  setSearchTerm,
   title,
 }: OnResultSearchClickType) => {
-  if (setSearch) setSearch(title);
+  if (setSearchTerm) setSearchTerm(title);
```
Comment (SharglutDev) [inline front/src/common/Map/utils.ts]: Why do we need to update this state if we close the popup anyway ?

File: front/src/styles/scss/common/map/_mapSearch.scss [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        font-size: 0.7rem;
+        color: var(--primary);
+      }
+      &.main {
+        .name {
+          font-size: 1.1rem;
+          padding-left: 0;
+          font-weight: 500;
+        }
+      }
```
Comment (SharglutDev) [inline front/src/styles/scss/common/map/_mapSearch.scss]: Sometimes the spacing between the trigram and the name is very short

Capture d’écran 2024-02-22 à 09 17 58

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for this great PR, search logic is much simpler and it works very well ! Left some comments.

**Review by flomonster — APPROVED** [review]
LGTM (only reviewed editoast)

File: front/src/styles/scss/common/map/_mapSearch.scss (hunk 4 above)
Comment (nicolaswurtz) [reply front/src/styles/scss/common/map/_mapSearch.scss]: We could use, according to your recommendation, monospace font.

File: front/src/common/Map/Search/MapSearchOperationalPoint.tsx (hunk 2 above)
Comment (nicolaswurtz) [reply front/src/common/Map/Search/MapSearchOperationalPoint.tsx]: We discussed about it, I think considering
- small size
- logic-less of the buttons
- no able to reuse (buttons change according to search object)

I think we can let thoses inside component. Thx anyway!

File: front/src/common/Map/utils.ts (hunk 3 above)
Comment (nicolaswurtz) [reply front/src/common/Map/utils.ts]: Yes, thx!

File: front/src/styles/scss/common/map/_mapSearch.scss (hunk 4 above)
Comment (nicolaswurtz) [reply front/src/styles/scss/common/map/_mapSearch.scss]: Done!

File: front/src/common/Map/Search/MapSearchOperationalPoint.tsx (hunk 1 above)
Comment (nicolaswurtz) [reply front/src/common/Map/Search/MapSearchOperationalPoint.tsx]: Yes, you're right! Done.

**Review by SharglutDev — APPROVED** [review]
Lgtm, good job !

**Pull request merged**